cpp/include/cuvs/neighbors/hnsw.h

```diff
@@ -0,0 +1,204 @@
+ *
+ * @param[in] index cuvsHnswIndex_t to allocate
+ * @return HnswError_t
+ */
+cuvsError_t cuvsHnswIndexCreate(cuvsHnswIndex_t* index);
+
+/**
+ * @brief De-allocate HNSW index
+ *
+ * @param[in] index cuvsHnswIndex_t to de-allocate
+ */
+cuvsError_t cuvsHnswIndexDestroy(cuvsHnswIndex_t index);
+
+/**
+ * @}
+ */
+
+/**
+ * @defgroup hnsw_c_index_search C API for CUDA ANN Graph-based nearest neighbor search
+ * @{
+ */
+/**
+ * @brief Search a HNSW index with a `DLManagedTensor` which has underlying
+ *        `DLDeviceType` equal to `kDLCPU`, `kDLCUDAHost`, or `kDLCUDAManaged`.
+ *        It is also important to note that the HNSW Index must have been built
+ *        with the same type of `queries`, such that `index.dtype.code ==
+ *        queries.dl_tensor.dtype.code`
+ *        Supported types for input are:
+ *        1. `queries`: `kDLDataType.code == kDLFloat` or `kDLDataType.code == kDLInt` and
+ * `kDLDataType.bits = 32`
+ *        2. `neighbors`: `kDLDataType.code == kDLUInt` and `kDLDataType.bits = 64`
+ *        3. `distances`: `kDLDataType.code == kDLFloat` and `kDLDataType.bits = 32`
+ *
+ * @code {.c}
+ * #include <cuvs/core/c_api.h>
+ * #include <cuvs/neighbors/hnsw.h>
+ *
+ * // Create cuvsResources_t
+ * cuvsResources_t res;
+ * cuvsError_t res_create_status = cuvsResourcesCreate(&res);
+ *
+ * // Assume a populated `DLManagedTensor` type here
+ * DLManagedTensor dataset;
+ * DLManagedTensor queries;
+ * DLManagedTensor neighbors;
+ *
+ * // Create default search params
+ * cuvsHnswSearchParams_t params;
+ * cuvsError_t params_create_status = cuvsHnswSearchParamsCreate(&params);
+ *
+ * // Search the `index` built using `cuvsHnswBuild`
+ * cuvsError_t search_status = cuvsHnswSearch(res, params, index, &queries, &neighbors,
+ * &distances);
+ *
+ * // de-allocate `params` and `res`
+ * cuvsError_t params_destroy_status = cuvsHnswSearchParamsDestroy(params);
+ * cuvsError_t res_destroy_status = cuvsResourcesDestroy(res);
+ * @endcode
+ *
+ * @param[in] res cuvsResources_t opaque C handle
+ * @param[in] params cuvsHnswSearchParams_t used to search Hnsw index
+ * @param[in] index cuvsHnswIndex which has been returned by `cuvsHnswBuild`
+ * @param[in] queries DLManagedTensor* queries dataset to search
+ * @param[out] neighbors DLManagedTensor* output `k` neighbors for queries
+ * @param[out] distances DLManagedTensor* output `k` distances for queries
+ */
+cuvsError_t cuvsHnswSearch(cuvsResources_t res,
+                           cuvsHnswSearchParams_t params,
+                           cuvsHnswIndex_t index,
+                           DLManagedTensor* queries,
+                           DLManagedTensor* neighbors,
+                           DLManagedTensor* distances);
+
+/**
+ * @}
+ */
+
+/**
+ * @defgroup hnsw_c_serialize HNSW C-API serialize functions
+ * @{
+ */
+
+/**
+ * Load hnswlib index from file which was serialized from a HNSW index.
+ * NOTE: The loaded hnswlib index is immutable.
```

Comment:
This is a start, but it still needs to clearly describe that this is meant to support a cagra graph that's been serialized to hnswlib format only through cuVS, since it's been patched to support this.